frontend/actions.js

```diff
@@ -98,58 +89,133 @@ export default ({setState, getState}) => {
     displayAboutOverlay: !state.displayAboutOverlay,
   })
 
-  const inputAddress = (state, e) => ({
-    sendAddress: e.target.value,
+  const validateSendAddress = (state) => setState({
+    sendAddress: Object.assign(state.sendAddress,
+      {validationMsg: !Cardano.isValidAddress(state.sendAddress.fieldValue) ? 'Invalid address' : ''}),
   })
 
-  // is being called debounced, thus we need to getState
-  // TODO maybe it'd be good practice to only use getState and avoid this problem?
+  const validateSendAmount = (state) => {
+    let validationMsg = ''
+    const amount = parseFloat(state.sendAmount.fieldValue)
+    if (isNaN(amount)) {
+      validationMsg = 'Invalid format: Amount has to be a number'
+    }
+    if (amount.toFixed(6) <= 0) {
+      validationMsg = 'Invalid format: Amount has to be a positive number'
+    }
+    if (amount.toFixed(6) > state.balance / 1000000) {
+      validationMsg = `Insufficient funds. Your balance is ${state.balance / 1000000} ADA.`
+    }
+    setState({sendAmount: Object.assign(state.sendAmount, {validationMsg})})
```

Comment:
..I think you've misunderstood Object.assign (and sorry for missing this in previous review)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

this would mutate state.sendAmount - what you've meant to do is Object.assign({}, state.sendAmount, {validationMsg})